File: example/u3v_camera2_opencv/u3v_camera2_opencv.cc
#include <opencv2/core.hpp>
#include <opencv2/imgproc.hpp>
#include <opencv2/highgui.hpp>

#include <ion/ion.h>

#include <exception>

using namespace ion;

#define FEATURE_GAIN_KEY "Gain"
#define FEATURE_EXPOSURE_KEY "ExposureTime"
#define NUM_BIT_SHIFT 0

// In this tutorial, we will create a simple application that obtains image data from a pair of usb3 vision sensors,
// and adds smoothing processing using OpenCV, and displays the data on the screen.

// Define parameters
//  Resize it according to the resolution of the sensor.
const int32_t width = 1920;
const int32_t height = 1080;
double gain = 400;
double exposure = 400;

int positive_pow(int base, int expo) {
    if (expo <= 0) {
        return 1;
    }
    if (expo == 1) {
        return base;
    } else {
        return base * positive_pow(base, expo-1);
    }
}


int main(int argc, char *argv[])
{
    try {
        // Define builders to build, compile, and execute pipelines.
        //  Build the pipeline by adding nodes to the Builder.
        Builder b;

        // Set the target hardware. The default is CPU.
        b.set_target(ion::get_host_target());

        // Load standard building block
        b.with_bb_module("ion-bb");

        //  Connect the input port to the Node instance created by b.add().
        Node n = b.add("image_io_u3v_cameraN_u16x2")(&gain, &exposure, &gain, &exposure)
            .set_param(
                Param("frame_sync", false),
                Param("gain_key", FEATURE_GAIN_KEY),
                Param("exposure_key", FEATURE_EXPOSURE_KEY),
                Param("realtime_display_mode", true),
                Param("enable_control", true)
                );

        // Map output buffer and ports by using Port::bind.
        // - output0: output 0 of the obtained video data
        // - output1: output 1 of the obtained video data
        // - frame_count: output of the frame number of the obtained video
        std::vector< int > buf_size = std::vector < int >{ width, height };
        Buffer<uint16_t> output0(buf_size);
        Buffer<uint16_t> output1(buf_size);
        Buffer<uint32_t> frame_count(1);

        n["output"].bind(std::vector<Buffer<uint16_t>>{output0, output1});
        n["frame_count"].bind(frame_count);

        // Obtain image data continuously for 100 frames to facilitate operation check.
        int loop_num = 100;
        int coef =  positive_pow(2, NUM_BIT_SHIFT);
        for (int i = 0; i < loop_num; ++i)
        {
            // JIT compilation and execution of pipelines with Builder.
            b.run();

            // Convert the retrieved buffer object to OpenCV buffer format.
            //  C and D are objects that store the result after smoothing.
            cv::Mat A(height, width, CV_16UC1, output0.data());
            cv::Mat B(height, width, CV_16UC1, output1.data());
            cv::Mat C(height, width, CV_16UC1);
            cv::Mat D(height, width, CV_16UC1);

            // Depends on sensor image pixel format, apply bit shift on images
            A = A * coef;
            B = B * coef;

            // Perform smoothing
            cv::medianBlur(A, C, 5);
            cv::medianBlur(B, D, 5);

            // Display the image
            cv::imshow("A", A);
            cv::imshow("B", B);
            cv::imshow("C", C);
            cv::imshow("D", D);

            // Wait for key input
            //   When any key is pressed, close the currently displayed image and proceed to the next frame.
            cv::waitKey(1);
        }

    } catch (const ion::Error& e) {
        std::cerr << e.what() << std::endl;
        return 1;
    }

 return 0;
}

File: example/u3v_camera2_opencv/u3v_camera2_opencv.py
from ionpy import Node, Builder, Buffer, PortMap, Port, Param, Type, TypeCode
import numpy as np
import cv2

from sys import platform

feature_gain_key = 'Gain'
feature_exposure_key = 'ExposureTime'
num_bit_shift = 0

if __name__ == "__main__":

    # Define parameters
    width = 1920
    height = 1080
    gain = 400
    exposure = 400

    # Build the pipeline by adding nodes to this builder.
    builder = Builder()

    #   Set the target hardware, The default is CPU.
    builder.set_target('host')

    #   Load building block module from the library
    builder.with_bb_module("ion-bb")


    # Define Input Port
    #    Port class would be  used to define dynamic O/O for each node.
    t = Type(TypeCode.Float, 64, 1)
    gain0_p = Port('gain0', t, 0)
    gain1_p = Port('gain1', t, 0)
    exposure0_p = Port('exposure0', t, 0)
    exposure1_p = Port('exposure1', t, 0)

    # Params
    num_devices = Param('num_devices', '2')
    frame_sync = Param('frame_sync', 'false')
    gain_key = Param('gain_key', feature_gain_key)
    exposure_key = Param('exposure_key', feature_exposure_key)
    realtime_diaplay_mode = Param('realtime_diaplay_mode', 'true')
    enable_control = Param('enable_control', 'true')

    #    Add node and connect the input port to the node instance
    node = builder.add('image_io_u3v_cameraN_u16x2')\
        .set_iports([gain0_p, gain1_p, exposure0_p, exposure1_p, ])\
        .set_params([num_devices, frame_sync, gain_key, exposure_key, realtime_diaplay_mode, enable_control])


    # Define Output Port
    out_p = node.get_port('output')
    frame_count_p = node.get_port('frame_count')

    # input values
    gain0_p.bind(gain)
    gain1_p.bind(gain)
    exposure0_p.bind(exposure)
    exposure1_p.bind(exposure)

    # output values
    odata0 = np.full((height, width), fill_value=0, dtype=np.uint16)
    output0 = Buffer(array=odata0)
    out_p[0].bind(output0)

    odata1 = np.full((height, width), fill_value=0, dtype=np.uint16)
    output1 = Buffer(array=odata1)
    out_p[1].bind(output1)

    fcdata = np.full((1), fill_value=0, dtype=np.uint32)
    frame_count = Buffer(array=fcdata)
    frame_count_p.bind(frame_count)

    loop_num = 100

    for x in range(loop_num):

        # running the builder
        builder.run()

        odata0 *= pow(2, num_bit_shift)
        odata1 *= pow(2, num_bit_shift)

        median_output0_np_HxW = cv2.medianBlur(odata0, 5)
        median_output1_np_HxW = cv2.medianBlur(odata1, 5)

        cv2.imshow("A", odata0)
        cv2.imshow("B", odata1)
        cv2.imshow("C", median_output0_np_HxW)
        cv2.imshow("D", median_output1_np_HxW)

        cv2.waitKey(1)

    cv2.destroyAllWindows()
